### PR TITLE
[enhancement] allow for ```onedal._backend``` modules to be imported directly

### PR DESCRIPTION
**Pull request by icfaust**
## Description

The dynamic setting of ```onedal._backend``` in onedal's ```__init__.py``` requires modifying sys.modules for proper operation.  Because these modules come from a shared object, reloading shouldn't be an issue.

Also removes ```_get_backend``` because of the obscurity it causes in understanding backend calls.

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.
- [x] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

**Comment by icfaust** [comment]
Rerun: http://intel-ci.intel.com/ef9fb29a-665e-f1f3-9f24-a4bf010d0e2d

File: onedal/_base.py [hunk 1]
```diff
@@ -16,23 +16,13 @@
...
-    if method:
-        return getattr(result, method)(*args, **kwargs)
-    return result
+from .common._policy import _get_policy
 
 
 class BaseEstimator(ABC):
-    def _get_backend(self, module, submodule=None, method=None, *args, **kwargs):
-        return _get_backend(_backend, module, submodule, method, *args, **kwargs)
+    _backend = onedal_backend
```
Comment (samir-nasibli) [inline onedal/_base.py:25]: The idea here was not to have a link to the whole backend  lib as a property to the estimators, now it is broken.

File: onedal/basic_statistics/basic_statistics.py [hunk 2]
```diff
@@ -19,13 +19,16 @@
...
-from ..common._base import BaseEstimator
+from .._backend import basic_statistics as onedal_backend
+from .._base import BaseEstimator
 from ..datatypes import _convert_to_supported, from_table, to_table
 from ..utils import _is_csr
 from ..utils.validation import _check_array
 
 
 class BaseBasicStatistics(BaseEstimator, metaclass=ABCMeta):
+    _backend = onedal_backend
```
Comment (samir-nasibli) [inline onedal/basic_statistics/basic_statistics.py:30]: If we already defined _backend in BaseEstimator, why it is needed here?

File: onedal/decomposition/pca.py [hunk 3]
```diff
@@ -21,7 +21,9 @@
 from sklearn.decomposition._pca import _infer_dimension
 from sklearn.utils.extmath import stable_cumsum
 
-from ..common._base import BaseEstimator
+import onedal._backend.decomposition.dim_reduction as onedal_backend
```
Comment (samir-nasibli) [inline onedal/decomposition/pca.py:24]: I think this naming a bit confusing

File: onedal/__init__.py [hunk 4]
```diff
@@ -49,12 +49,22 @@
...
 if _is_dpc_backend:
     try:
         import onedal._onedal_py_spmd_dpc as _spmd_backend
 
+        for mod in sys.modules.copy():
+            if mod.startswith(_spmd_backend.__name__):
+                sys.modules[
+                    mod.replace(_spmd_backend.__name__, "onedal._spmd_backend")
+                ] = sys.modules[mod]
+
```
Comment (samir-nasibli) [inline onedal/__init__.py:67]: Please add comment in code base why is this needed.

File: onedal/_base.py (hunk 1 above)
Comment (icfaust) [reply onedal/_base.py:25]: I guess we can discuss this, ideally we would alias in the necessary backend functions as part of the estimators in a case-by-case basis. i.e. for basic statistics we would do:
```python
from onedal._backend.basic_statistics import compute

class BasicStatistics(BaseEstimator, metaclass=ABCmeta):
      _compute = compute
```

Comment (icfaust) [reply onedal/_base.py:25]: This would be very clear to the developer and the user where and what is being used, while also maintaining spmd inheritance simple.

File: onedal/basic_statistics/basic_statistics.py (hunk 2 above)
Comment (icfaust) [reply onedal/basic_statistics/basic_statistics.py:30]: @samir-nasibli  ```_backend``` in this case is ```onedal._backend.basic_statistics```.  In a future PR we can assign the necessary methods to the estimators as functions, and overwrite them in the spmd estimators in a case-by-case basis.  That is a lot of work to switch this in this PR, but the correct one for clarity and extensibility. The changes here are an intermediate step. Otherwise this PR may be a 500+ line change.

File: onedal/__init__.py (hunk 4 above)
Comment (icfaust) [reply onedal/__init__.py:75]: Done:
https://github.com/icfaust/scikit-learn-intelex/blob/dev/remove_get_backend/onedal/__init__.py#L52

Comment (samir-nasibli) [reply onedal/__init__.py:75]: Thank you, @icfaust! I understand the idea now. Could we address this through a test, assuming we already have a reproducer for the issue?

File: onedal/cluster/dbscan.py [hunk 5]
```diff
@@ -16,15 +16,18 @@
...
 
-from ..common._base import BaseEstimator
+from .._base import BaseEstimator
 from ..common._mixin import ClusterMixin
 from ..datatypes import _convert_to_supported, from_table, to_table
 from ..utils import _check_array
 
 
 class BaseDBSCAN(BaseEstimator, ClusterMixin):
+    _backend = onedal_backend
```
Comment (samir-nasibli) [inline onedal/cluster/dbscan.py:29]:
```suggestion
    _backend = onedal_dbscan_clustering
```

File: onedal/cluster/dbscan.py [hunk 6]
```diff
@@ -16,15 +16,18 @@
 
 import numpy as np
 
+import onedal._backend.dbscan.clustering as onedal_backend
```
Comment (samir-nasibli) [inline onedal/cluster/dbscan.py:19]:
```suggestion
import onedal._backend.dbscan.clustering as onedal_dbscan_clustering
```

or 

```suggestion
import onedal._backend.dbscan.clustering as onedal_backend_dbscan_clustering
```

I believe this makes better the readability of the code.

**Review by samir-nasibli — COMMENTED** [review]
suggestions on the naming

**Comment by ethanglaser** [comment]
Might be worth setting up a call to discuss this. Not sure how others feel but I am having a hard time understanding the changes.

**Comment by icfaust** [comment]
> Might be worth setting up a call to discuss this. Not sure how others feel but I am having a hard time understanding the changes.

Fair point, I will move this to draft for the time being and will create a meeting.
